### ISSUE TITLE
Decorators with Func<ServiceType> parameters wipe the dependency graph

### ISSUE THREAD

**Issue by fsateler** [simpleinjector/SimpleInjector#414]
When using a decorator with a [`Func<T>` constructor argument](https://simpleinjector.readthedocs.io/en/latest/aop.html#decorators-with-func-t-decoratee-factories), the dependency graph from that point onward becomes invisible to the user:

```csharp
// code for linqpad
interface IService { }
class MyService : IService { }

class MyServiceDecorator : IService {
    private readonly Func<IService> DecorateeFactory;
	public MyServiceDecorator(Func<IService> factory) {
	    DecorateeFactory = factory;
	}
}

void Main()
{
	var container = new Container();
	
	container.Register<IService, MyService>();
	container.RegisterDecorator<IService, MyServiceDecorator>();
	container.Verify();
	
	container.GetCurrentRegistrations().Select(r => r.VisualizeObjectGraph()).Dump();
}
```
The current registrations have a single element, which is described as follows:

```
UserQuery.MyServiceDecorator(
    Func<UserQuery.IService>())
```

This means that any decorators, and the real implementation type is invisible. In turn, this makes verifying properties of the object graph impossible if such decorators are used (at least in my use case, these decorators come last in the registration order, and thus the entire object graph is wiped).

Tested with versions 3.3.2, 4.0.0 and 4.0.3 .

**Comment by dotnetjunkie** [comment]
This is kinda by-design, or at least, me and @qujck discussed this a few years back, but we decided not to 'fix' this, because:
 1. this is kind-of the correct behavior, because you're injecting a singleton factory -not the graph itself.
 2. It's actually not easy at all to implement it.
 3. Since `VisualizeObjectGraph` uses the same structure (i.e. `KnownDependencies`) that the diagnostic sub system uses, changing the graph will cause the diagnostic sub system to incorrectly warn about Lifestyle Mismatches, because now it might seem that a singleton decorator depends on a non-singleton dependency, while in reality it depends on a factory with a singleton lifestyle (which means everything is fine).

We did realize however that this disallows correctly analyzing object graphs when iterating the `KnownDependencies` or vizualizing the graph using `VisualizeObjectGraph`.

**Comment by fsateler** [comment]
Hi Steven, thanks for the quick reply. The arguments you point out make this a difficult problem. In particular, warning about lifestyle mismatches would be a no-no for final users, and I imagine having two separate analyzer infrastructures is unpleasant to say the least. However, I think there can be a solution here. 

Whenever such a `Func<T>` is generated by SimpleInjector[1] when registering a decorator, SimpleInjector could also register a dummy type, `SimpleInjectorFuncProxy<T>`, which would take in as parameters the concrete class the `Func<T>` would return. This proxy would have a transient lifestyle, and have no consumers, and thus it would have no lifestyle mismatches. Then one could look for these types and perform the analysis on it.

Pros of this approach: it solves my problem 😄 
Cons of this approach:
- A slightly larger registration list, depending on the amount of automatically-generated `Func<T>` are created. 
- Possible diagnostic messages due to "container-registered" types being in use. Could be mitigated by explicit registrations.
- Difficulty of implementation? I have no idea if implementing this would be hard.

[1] I certainly don't think this extends to functions passed by the user.

**Comment by dotnetjunkie** [comment]
Can you give an example of what the output of `VisualizeObjectGraph` should be in your case?

**Comment by fsateler** [comment]
I'm going to extend the example a bit to make this clearer:

```csharp
interface IService {
}
class MyService : IService {
}

class MyServiceDecorator1 : IService {
    private readonly IService Decoratee;
	public MyServiceDecorator1(IService decoratee) {
	    Decoratee = decoratee;
	}
}
class MyServiceDecorator2 : IService {
    private readonly Func<IService> DecorateeFactory;
	public MyServiceDecorator2(Func<IService> factory) {
	    DecorateeFactory = factory;
	}
}

void Main()
{

	var container = new Container();
	
	container.Register<IService, MyService>();
	container.RegisterDecorator<IService, MyServiceDecorator1>();
	container.RegisterDecorator<IService, MyServiceDecorator2>();
	container.Verify();
	
	container.GetCurrentRegistrations().Select(r => r.VisualizeObjectGraph()).Dump();
}
```

Should output this:

```
UserQuery.MyServiceDecorator2(
    Func<UserQuery.IService>())
---
SimpleInjectorFuncProxy<UserQuery.IService>(
    UserQuery.MyServiceDecorator1(
        UserQuery.MyService()))
```

**Comment by dotnetjunkie** [comment]
I don't like having a proxy in there, but I think you actually provided a reasonable and implementable solution, which is: let the decoratee show up as individual registration in the list of _current registrations_. In fact, I think that from a diagnostic perspective, I think this `InstanceProducer` already exists, since the bottom part of the graph (dec1 + service) is analyzed as well.

This means only thing that has to happen is to include this InstanceProducer in the list. It would result in the following output:

```
UserQuery.MyServiceDecorator2(
    Func<UserQuery.IService>())
---
UserQuery.MyServiceDecorator1(
    UserQuery.MyService())
```

Would that solve your problem?

**Comment by fsateler** [comment]
That would mean `MyServiceDecorator1` gets an implicit concrete registration (that is, `ServiceType` would be `MyServiceDecorator` instead of `IService`)? That would make analysis a bit harder, but I guess it will work too.

**Comment by dotnetjunkie** [comment]
Not at all. ServiceType can still be IService. There is no reason for that information to get lost. InstanceProducer instances returned from GetCurrentRegistrations don't have to be resolvable through GetInstance. In fact *all* InstanceProducers that are related to a container instance will be returned (unless GCed before).

**Comment by fsateler** [comment]
Unless I misunderstood, it is not really *all* InstanceProducers that are returned, as there should not be an InstanceProducer for (undecorated) `UserQuery.MyService` in the list. Or are you suggesting the actual list would be:

```
UserQuery.MyServiceDecorator2(
    Func<UserQuery.IService>())
---
UserQuery.MyServiceDecorator1(
    UserQuery.MyService())
---
UserQuery.MyService()
```

?

**Comment by dotnetjunkie** [comment]
No, you are correct. If a service has 4 decorators, it will still end up as one item in the list, because a decorator is just an enrichment of that registration. But imagine this:

``` c#
var a1 = Lifestyle.Transient.CreateProducer<IService, ServiceA>(container);
var a2 = Lifestyle.Transient.CreateProducer<IService, ServiceA>(container);
var b1 = Lifestyle.Transient.CreateProducer<IService, ServiceB>(container);
var b2 = Lifestyle.Transient.CreateProducer<IService, ServiceB>(container);

container.GetCurrentRegistrations();
```

All 4 `InstanceProducer` instances will be returned from `GetCurrentRegistrations()`.

When it comes to decoratee factories however, things get a bit different. In that case, since creation of the decoratee is postponed by the factory, the decoratee itself becomes a new root type, as in: it is resolved directly from the container, not injected into anything else.

From that perspective it makes sense to see the decoratee as its own registration. This would yield in something similar:

``` c#
container.Register<IService, MyService>();
container.RegisterDecorator<IService, MyServiceDecorator1>();

container.Register<MyServiceDecorator2>(
    () => new MyServiceDecorator2(
        () => container.GetInstance<IService>()));
```

This will result in two items in the `GetCurrentRegistrations` list: one for IService/MyService, and one for MyServiceDecorator2.

**Comment by dotnetjunkie** [comment]
The following test reproduces the issue:

``` c#
// Add to DecoratorTests.cs
// Feature #414.
[TestMethod]
public void GetCurrentRegistrations_DecoratorWithFuncDecoratee_RegistrationsContainsRegistrationForDecorateeAsWell()
{
    // Act
    var container = new Container();

    container.Register<INonGenericService, RealNonGenericService>();
    container.RegisterDecorator<INonGenericService, NonGenericServiceDecorator>();
    container.RegisterDecorator<INonGenericService, NonGenericServiceDecoratorWithFunc>();

    container.Verify();

    // Act
    var registrations = container.GetCurrentRegistrations();

    // Assert
    var graphs = (
        from registration in registrations
        where registration.ServiceType == typeof(INonGenericService)
        let graph = registration.VisualizeObjectGraph()
        orderby graph
        select graph)
        .ToArray();

    Assert.AreEqual(2, graphs.Length,
        $"Two graphs expected, but {graphs.Length} found. Graphs:\n{string.Join(Environment.NewLine, graphs)}.");

    // This is the registration for the outer most (Func<T>) decorator.
    Assert.AreEqual(
        expected: $"{nameof(NonGenericServiceDecoratorWithFunc)}( Func<{nameof(INonGenericService)}>())", 
        actual: graphs[1].TrimInside());

    // This is the registration that contains the graph of part that the Func<T> captures.
    Assert.AreEqual(
        expected: $"{nameof(NonGenericServiceDecorator)}( {nameof(RealNonGenericService)}())",
        actual: graphs[0].TrimInside());
}
```

**Comment by dotnetjunkie** [comment]
There has been some fortunate development concerning a quite related bug (#759) concerning the visualization of collections (which are another form of factories). This has been reported with @pavel-pargachev and discussed with both @pavel-pargachev and @kwlin.

The solution that has now been implemented with #759 is to [differentiate `KnownRelationship` by their type](https://github.com/simpleinjector/SimpleInjector/commit/926563f239efd67f22a0c09ead94a0c9767269a1#diff-2b4b36ad7f928c99fef28b494733c30cR69). #759 now allows Simple Injector to internally flag a relationship for being "visualization only", which might be the solution for this problem.

I know it has been a while since this thread was active, @fsateler, but I would like to suggest another way of visualizing object graphs that consist of `Func<T>` decorators. Considering your initial object graph, I suggest letting Simple Injector visualize the graph as follows:

```
MyServiceDecorator2(
    Func<IService>(
        MyServiceDecorator1(
            MyService())))
```

Or, with lifestyles (which is currently the default):

```
MyServiceDecorator2( // Transient
    Func<IService>( // Singleton
        MyServiceDecorator1( // Transient
            MyService()))) // Transient
```

I think this is the best solution, because it gives you, as a user, visually exactly what you want to see, without seeing a broken object graph and without trying to query to get the whole object graph visible, while still seeing the factory in the middle, which is still an important detail when viewing the graph.

Feedback is welcome.

**Comment by fsateler** [comment]
Hi!

Unfortunately, I have since moved on, and I don't remember all the details. That said, the proposed visualization looks good to me.

**Comment by dotnetjunkie** [comment]
This bug has been fixed in [v4.9.1](https://github.com/simpleinjector/SimpleInjector/releases/tag/v4.9.1).
